### PR TITLE
Fix NaN/inf propagation in scipy.linalg.toeplitz and hankel

### LINKED ISSUE #38635: jax.scipy.linalg.hankel incorrectly propagates row[0] NaN into output
### Description

`jax.scipy.linalg.hankel` appears to incorrectly propagate `row[0]` into the output when `row[0]` is `NaN`.

For a Hankel matrix, the first column should come from `column`, and the last row should come from `row`. As in SciPy's `scipy.linalg.hankel`, `row[0]` should be ignored because it overlaps with `column[-1]`.

In the example below, `row[0]` is `NaN`, but it should not contaminate the top row. The expected result should use values from `column` followed by `row[1:]`. However, JAX returns additional `NaN` values where `inf` and `-0.0` are expected.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import warnings

warnings.filterwarnings("ignore")
os.environ.setdefault("JAX_PLATFORMS", "cpu")

try:
    import numpy as np
    import jax.numpy as jnp
    import jax.scipy.linalg as jlinalg
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

column = np.array([np.nan, np.inf, -0.0], dtype=np.float32)
row = np.array([np.nan, 2.0, 3.0], dtype=np.float32)

expected = np.array(
    [
        [np.nan, np.inf, -0.0],
        [np.inf, -0.0, 2.0],
        [-0.0, 2.0, 3.0],
    ],
    dtype=np.float32,
)

out = np.asarray(jlinalg.hankel(jnp.asarray(column), jnp.asarray(row)))

print("jax:", out.tolist())
print("expected:", expected.tolist())

if not np.array_equal(out, expected, equal_nan=True):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
jax: [[nan, nan, nan], [inf, nan, nan], [0.0, 2.0, 3.0]]
expected: [[nan, inf, -0.0], [inf, -0.0, 2.0], [-0.0, 2.0, 3.0]]
```

### Expected result

```
[[nan, inf, -0.0],
 [inf, -0.0, 2.0],
 [-0.0, 2.0, 3.0]]
```

The `NaN` at `row[0]` should not be propagated into the constructed Hankel matrix. The matrix should be constructed from:

```
[column[0], column[1], column[2], row[1], row[2]]
```

so the effective sequence should be:

```
[nan, inf, -0.0, 2.0, 3.0]
```

JAX instead appears to let `row[0]` affect positions where `column[1]` and `column[2]` should be used.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### LINKED ISSUE #38636: jax.scipy.linalg.toeplitz incorrectly propagates NaN to the entire output matrix
### Description

`jax.scipy.linalg.toeplitz` appears to incorrectly propagate `NaN` values when constructing a Toeplitz matrix from inputs containing nonfinite values.

A Toeplitz matrix should be constructed by selecting entries from the first column and first row. In this example, `row[0]` is `NaN`, but it should only affect positions that actually correspond to the overlapping diagonal value. It should not contaminate the whole output matrix.

However, JAX returns an all-`NaN` matrix, while the expected output contains finite values, `inf`, and `-0.0` in positions selected from `column` and `row`.

This looks like the implementation may be using arithmetic or masking in a way that allows `NaN` to propagate, instead of performing pure indexing/selection.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import warnings

warnings.filterwarnings("ignore")
os.environ.setdefault("JAX_PLATFORMS", "cpu")

try:
    import numpy as np
    import jax.numpy as jnp
    import jax.scipy.linalg as jlinalg
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

column = np.array([np.nan, np.inf, -0.0], dtype=np.float32)
row = np.array([np.nan, 2.0, 3.0], dtype=np.float32)

expected = np.array(
    [
        [np.nan, 2.0, 3.0],
        [np.inf, np.nan, 2.0],
        [-0.0, np.inf, np.nan],
    ],
    dtype=np.float32,
)

out = np.asarray(jlinalg.toeplitz(jnp.asarray(column), jnp.asarray(row)))

print("jax:", out.tolist())
print("expected:", expected.tolist())

if not np.array_equal(out, expected, equal_nan=True):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
jax: [[nan, nan, nan], [nan, nan, nan], [nan, nan, nan]]
expected: [[nan, 2.0, 3.0], [inf, nan, 2.0], [-0.0, inf, nan]]
```

### Expected result

```
[[nan, 2.0, 3.0],
 [inf, nan, 2.0],
 [-0.0, inf, nan]]
```

For a Toeplitz matrix, the output should be formed by selecting values from `column` and `row`. `NaN` values should only appear in positions that correspond to an actual selected `NaN` input element. They should not propagate through unrelated finite, infinite, or signed-zero entries.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### PR DESCRIPTION
**Pull request by GodlyDonuts**
### What

`jax.scipy.linalg.toeplitz` and `jax.scipy.linalg.hankel` return an all-`NaN` matrix whenever any input element is non-finite, even for output cells that should select a perfectly finite value.

```python
>>> import jax.numpy as jnp, jax.scipy.linalg as jsl
>>> c = jnp.array([jnp.nan, jnp.inf, -0.0], dtype=jnp.float32)
>>> r = jnp.array([jnp.nan, 2.0, 3.0], dtype=jnp.float32)
>>> jsl.toeplitz(c, r)
Array([[nan, nan, nan],
       [nan, nan, nan],
       [nan, nan, nan]], dtype=float32)   # every scipy/numpy entry here is finite except the diagonal
```

SciPy returns `[[nan, 2, 3], [inf, nan, 2], [-0, inf, nan]]`; `hankel` is affected the same way.

### Why

Both `_toeplitz` and `_hankel` materialize their entries with `lax.conv_general_dilated_patches`, i.e. an im2col implemented as a convolution against a 0/1 kernel. Convolution is a multiply-accumulate, and `0 * inf == nan`, `0 * nan == nan`. So in each output cell's dot product the masked-out terms aren't actually zero — a single non-finite input poisons the entire result, regardless of which entry that cell was supposed to select.

### Fix

A Toeplitz/Hankel matrix is a pure *selection* of the input values, so construct it by indexing rather than arithmetic. Each layout has a closed-form index into the flattened diagonals:

```
toeplitz: A[i, j] = elems[ncols - 1 + j - i],  elems = concat(c[::-1], r[1:])
hankel:   H[i, j] = v[i + j],                   v     = concat(c, r[1:])
```

Gathering with `broadcasted_iota` index grids touches the data only by selection: non-finite values stay confined to their own cells and signed zeros survive intact. As a bonus it removes the convolution + `flip`, which is both simpler and cheaper than im2col for this access pattern. The `@jnp.vectorize` batching wrappers are unchanged.

### Tests

- New `testToeplitzNonFiniteValues` / `testHankelNonFiniteValues` lock in the SciPy-matching output (including `-0.0` sign preservation).
- The existing `testToeplitzConstruction` / `testHankelConstruction` suites pass; I also checked the new indexing against `scipy.linalg` across every column/row shape combination they cover.

Fixes #38636
Fixes #38635

### CODE REVIEW

**Comment by GodlyDonuts** [comment]
Hi @jakevdp, would you be able to review this PR when you have a chance, or route it to the appropriate reviewer? It fixes unintended NaN/inf propagation in `jax.scipy.linalg.toeplitz` and `hankel` and adds regression coverage. Happy to make any changes needed.

**Comment by jakevdp** [comment]
This falls under #38651 – it may take a while for us to get to it.
